src/Microsoft.AspNetCore.Razor.Language/Legacy/HtmlMarkupParser.cs

```diff
@@ -222,282 +259,466 @@ protected void SkipToAndParseCode(Func<SyntaxToken, bool> condition)
             }
         }
 
-        protected static Func<SyntaxToken, bool> IsSpacingToken(bool includeNewLines)
-        {
-            return token => token.Kind == SyntaxKind.Whitespace || (includeNewLines && token.Kind == SyntaxKind.NewLine);
-        }
-
-        private void OtherParserBlock()
-        {
-            AddMarkerTokenIfNecessary();
-            Output(SpanKindInternal.Markup);
-
-            using (PushSpanConfig())
-            {
-                CodeParser.ParseBlock();
-            }
-
-            Span.Start = CurrentLocation;
-            Initialize(Span);
-            NextToken();
-        }
-
-        private bool IsBangEscape(int lookahead)
-        {
-            var potentialBang = Lookahead(lookahead);
-
-            if (potentialBang != null &&
-                potentialBang.Kind == SyntaxKind.Bang)
-            {
-                var afterBang = Lookahead(lookahead + 1);
-
-                return afterBang != null &&
-                    afterBang.Kind == SyntaxKind.Text &&
-                    !string.Equals(afterBang.Content, "DOCTYPE", StringComparison.OrdinalIgnoreCase);
-            }
-
-            return false;
-        }
-
-        private void OptionalBangEscape()
+        /// <summary>
+        /// Reads the content of a tag (if present) in the MarkupDocument (or MarkupSection) context,
+        /// where we don't care about maintaining a stack of tags.
+        /// </summary>
+        private void ParseTagInDocumentContext(in SyntaxListBuilder<RazorSyntaxNode> builder)
```

Comment:
Should factor this method a lot like you have ParseCodeBlock in the CSharpCodeParser. i.e.
- if HtmlComment => HtmlComment()
- If XmlTag => XmlTag()
- If SpecialTag => `SpecialTag()`
- else `ParseHtmlStartTag()`

I'm sure I didn't capture everything but that's the gist. Feel free to add this to the feedback look. As it is right now it's spaghetti 😢